libs/core/kiln_ai/datamodel/basemodel.py

```diff
@@ -71,6 +68,118 @@ def string_to_valid_name(name: str) -> str:
 
 
+class KilnAttachmentModel(BaseModel):
+    path: Path = Field(description="The path to the attachment")
+
+    is_persisted: bool = Field(
+        default=False,
+        exclude=True,
+        description="Whether the attachment is persisted to its permanent location on disk. This is set automatically when the attachment is loaded from a file.",
+    )
+
+    @model_validator(mode="after")
+    def check_file_exists(self, info: ValidationInfo) -> Self:
+        context = info.context or {}
+        self.is_persisted = False
+        if context.get("loading_from_file", False):
+            # if we load from file, we assume it is persisted correctly
+            self.is_persisted = True
+            return self
+        if isinstance(self.path, str):
+            self.path = Path(self.path)
+        if not os.path.exists(self.path):
+            raise ValueError(f"File does not exist: {self.path}")
+        if not os.path.isfile(self.path):
+            raise ValueError(f"Path is not a file: {self.path}")
+        return self
+
+    @model_serializer
+    def serialize(self, info: SerializationInfo) -> dict[str, Path] | None:
+        # when the attachment is optional on the model, we get None here
+        if self is None:
+            return None
+
+        context = info.context or {}
+
+        # serialization may also be called by other parts of the system, the callers should
+        # explicitly set save_attachments to True if they want to save attachments
+        save_attachments: bool = context.get("save_attachments", False)
+        if not save_attachments:
+            return {"path": self.path}
+
+        dest_path: Path | None = context.get("dest_path", None)
+        if not dest_path or not isinstance(dest_path, Path):
+            raise ValueError(
+                f"dest_path must be a valid Path object when saving attachments, got: {dest_path}"
+            )
+        if not dest_path.is_dir():
+            raise ValueError("dest_path must be a directory when saving attachments")
+
+        # the attachment is already in the parent folder, so we don't need to copy it
+        # if the path is already relative, we consider it has been copied already
+        if self.is_persisted:
+            return {"path": self.path}
+
+        # copy file and update the path to be relative to the dest_path
+        path_before_copy = self.path
+        new_path = self.copy_file_to(dest_path)
+
+        # check if source was a temp file and clean it up
+        if KilnAttachmentModel.is_temp_file(path_before_copy):
+            try:
+                path_before_copy.unlink()
+            except (OSError, FileNotFoundError):
+                pass  # file is already gone, that's fine
+
+        self.path = new_path.relative_to(dest_path)
+        self.is_persisted = True
+
+        return {"path": self.path}
+
+    def copy_file_to(self, dest_folder: Path) -> Path:
+        filename = f"{str(uuid.uuid4().int)[:12]}{self.path.suffix}"
```

Comment:
P2: filename would be nice if it was the attachmetns name. IE if the attachmetn was `extraction: KilnAttachment` it would be "extraction.md". We'll have models with several attachments.

Can we just pass it in context?